### PR TITLE
Add Enclave Simulator for Internal Testing

### PR DESCRIPTION
**Pull request by karinazhou**
- Added support for enclave simulator bypassing the actual attestation.
- Added `BuildSimulator` property to enable the simulator support.

How to use the enclave simulator: 

Given `Attestation Protocol = SIM` and `Enclave Attestation Url= dummyURL` in the connection string and run the AE enclave-enabled tests.

NOTE:  This change is for internal testing only. The simulator support should be disabled in the official nuget package releases.

### CODE REVIEW

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 1]
```diff
@@ -3017,23 +3017,41 @@ private bool TryProcessFeatureExtAck(TdsParserStateObject stateObj)
...
         private bool IsValidAttestationProtocol(SqlConnectionAttestationProtocol attestationProtocol, string enclaveType)
         {
-            switch (enclaveType)
+            switch (enclaveType.ToUpper())
             {
                 case TdsEnums.ENCLAVE_TYPE_VBS:
                     if (attestationProtocol != SqlConnectionAttestationProtocol.AAS
+#if ENCLAVE_SIMULATOR
+                        && attestationProtocol != SqlConnectionAttestationProtocol.HGS
+                        && attestationProtocol != SqlConnectionAttestationProtocol.SIM)
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: Nit: If you put SIM above HGS, you can eliminate a line of code.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 2]
```diff
@@ -3485,23 +3485,41 @@ private bool TryProcessFeatureExtAck(TdsParserStateObject stateObj)
...
         private bool IsValidAttestationProtocol(SqlConnectionAttestationProtocol attestationProtocol, string enclaveType)
         {
-            switch (enclaveType)
+            switch (enclaveType.ToUpper())
             {
                 case TdsEnums.ENCLAVE_TYPE_VBS:
                     if (attestationProtocol != SqlConnectionAttestationProtocol.AAS
+#if ENCLAVE_SIMULATOR
+                        && attestationProtocol != SqlConnectionAttestationProtocol.HGS
+                        && attestationProtocol != SqlConnectionAttestationProtocol.SIM)
```
Comment (David-Engel) [inline src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs]: Nit: If you put SIM above HGS, you can eliminate a line of code.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 1 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: Good point. I will move SIM a line up.

File: src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs [hunk 3]
```diff
@@ -113,6 +113,11 @@ public enum SqlConnectionAttestationProtocol
         /// <include file='..\..\..\..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionAttestationProtocol.xml' path='docs/members[@name="SqlConnectionAttestationProtocol"]/AAS/*' />
         AAS = 1,
 
+#if ENCLAVE_SIMULATOR
+        /// <include file='..\..\..\..\doc\snippets\Microsoft.Data.SqlClient\SqlConnectionAttestationProtocol.xml' path='docs/members[@name="SqlConnectionAttestationProtocol"]/SIM/*' />
+        SIM = 2,
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs]: Is it okay to reserve enum value 2 for SIM?
Does this enum correspond to SQL Server's response as well?

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs [hunk 4]
```diff
@@ -3017,23 +3017,41 @@ private bool TryProcessFeatureExtAck(TdsParserStateObject stateObj)
...
+#endif
                     {
                         return false;
                     }
                     break;
 
                 case TdsEnums.ENCLAVE_TYPE_SGX:
+#if ENCLAVE_SIMULATOR
+                    if (attestationProtocol != SqlConnectionAttestationProtocol.AAS
+                        && attestationProtocol != SqlConnectionAttestationProtocol.SIM)
```
Comment (cheenamalhotra) [inline src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: I guess same can be done here, as David commented above.

File: src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs (hunk 3 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs]: This protocol number is defined in SimulatorEnclaveProviders.cs. It is used to create the SqlEnclaveAttestationParameters. I think we need this number to communicate with SQL Server to get a response.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 4 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: This one is a bit different. If I remove the `#else` branch, I will need to put `)` in a separate line which is a bit weird.

File: src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 2 above)
Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netfx/src/Microsoft/Data/SqlClient/TdsParser.cs]: Changed.

File: src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs (hunk 4 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/TdsParser.cs]: Fair enough!

File: src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs (hunk 3 above)
Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs]: My doubt was is it us who own 1 and 3 enum values for supported types (would also answer why 1 and 3 are for AAS/VBS), and if assigning 2 for SIM is also driver specific only?

Does server not care if we assign it 2 or 5 or any other value?

Comment (karinazhou) [reply src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs]: I don't think we can randomly pick other numbers as these protocolIDs (AAS / HGS). The protocolID is part of the attestationParameters which the driver needs to be sent to the server through sp_describe_parameter_encryption RPC call. It should be the same numbers across all the drivers targeting the same SQL Server.

Comment (cheenamalhotra) [reply src/Microsoft.Data.SqlClient/netcore/ref/Microsoft.Data.SqlClient.NetCoreApp.cs]: Resolving based on Offline discussion.

**Pull request merged**